src/components/VmStatus/getVmStatus.js

```diff
@@ -156,17 +131,138 @@ const isWaitingForVmi = vm => {
   return NOT_HANDLED;
 };
 
-export const getVmStatusDetail = (vm, launcherPod, importerPods, migration) =>
-  isBeingMigrated(vm, migration) || // must be precceding isRunning() since vm.status.ready is true for a migrating VM
+const getPodError = cdiPod => {
+  if (cdiPod) {
+    if (!isSchedulable(cdiPod)) {
+      return 'Pod scheduling failed.';
+    }
+
+    const failingContainer = getFailingContainerStatus(cdiPod);
+    if (failingContainer) {
+      return getContainerStatusReason(failingContainer);
+    }
+  }
+  return NOT_HANDLED;
+};
+
+const getDiskErrorMessage = (dVolume, diskName) => {
+  const dvSource = getDataVolumeSourceType(dVolume);
+  let diskAction;
+  switch (get(dvSource, 'type')) {
+    case DATA_VOLUME_SOURCE_PVC:
+      diskAction = 'Cloning';
+      break;
+    case DATA_VOLUME_SOURCE_UPLOAD:
+      diskAction = 'Uploading';
+      break;
+    case DATA_VOLUME_SOURCE_URL:
+      diskAction = 'Importing';
+      break;
+    default:
+      break;
+  }
+  return diskAction ? `${diskAction} disk ${diskName} failed` : `Disk ${diskName} failed`;
+};
+
+export const isPreparingDisks = (vm, cdiPods, dataVolumes) => {
+  const diskStatuses = [];
+  if (!dataVolumes) {
+    return NOT_HANDLED;
+  }
+  const dataVolumeVolumes = getVolumes(vm).filter(v => v.dataVolume);
+  dataVolumeVolumes.forEach(volume => {
+    const dVolume = dataVolumes.find(
+      dv => getName(dv) === volume.dataVolume.name && getNamespace(dv) === getNamespace(vm)
+    );
+    if (!dVolume) {
+      return;
+    }
+    const pod = findCdiPod(volume, cdiPods);
+    const podError = getPodError(pod);
+    const diskStatus = {
+      diskName: volume.name,
+      pod,
+    };
+    if (podError) {
+      diskStatuses.push({
+        ...diskStatus,
+        status: VM_STATUS_DISKS_FAILED,
+        message: getDiskErrorMessage(dVolume, volume.name),
+        verboseMessage: podError,
+      });
+    } else {
+      switch (get(dVolume, 'status.phase')) {
```

Comment:
can you please make this less verbose?

almost all the cases differ only by diskStatus string which could be resolved by function or object mappings

Reply:
I tried to make it less verbose. Please check again